frontends/p4/simplifyDefUse.cpp

```diff
@@ -339,6 +574,83 @@ class FindUninitialized : public Inspector {
         return loc;
     }
 
+    // Used in header-to-header and struct-to-struct assignments
+    void processHeadersInAssignment(const StorageLocation* dst, const StorageLocation* src) {
+        if (!dst || !src)
+            return;
+
+        auto dst_struct_storage = dst->to<StructLocation>();
+        auto src_struct_storage = src->to<StructLocation>();
+
+        if (!dst_struct_storage || !src_struct_storage)
+            return;
+
+        if (dst_struct_storage->isHeader() && src_struct_storage->isHeader()) {
+            auto val = headerDefs->find(src);
+            if (val)
+                headerDefs->update(dst, *val);
+            else
+                headerDefs->remove(dst);
+
+            return;
+        }
+
+        if (dst_struct_storage->isStruct() && src_struct_storage->isStruct()) {
+            auto dst_fields = dst_struct_storage->fields();
+            auto src_fields = src_struct_storage->fields();
+
+            auto it1 = dst_fields.begin();
+            auto it2 = src_fields.begin();
+            while (it1 != dst_fields.end() && it2 != src_fields.end()) {
+                processHeadersInAssignment(*it1, *it2);
+                ++it1;
+                ++it2;
+            }
+        }
+    }
+
+    // Used for more complex assignments
+    void processHeadersInAssignment(const StorageLocation* dst, const IR::Expression* src,
+                                    const IR::Type* src_type) {
+        if (!dst || !src || !src_type)
+            return;
+
+        if (auto dst_struct_storage = dst->to<StructLocation>()) {
+            if (dst_struct_storage->isHeader()) {
+                if (src->is<IR::StructExpression>()) {
+                    // always sets the valid bit to true
+                    headerDefs->update(dst, true);
+                } else if (src->is<IR::MethodCallExpression>()) {
+                    // return value of a function
+                    headerDefs->update(dst, true);
+                } else if (src->is<IR::ArrayIndex>()) {
+                    // TODO: header stacks
```

Comment:
Can't you just keep track of a valid bit for each slot of the array?
If the index is not constant you can assume the "or" of all valid bits of all elements for reads.
You can do this in a subsequent PR.

Reply:
I didn't want to put everything in a single PR because I thought it would make the review harder. I agree with your idea, it looks good to me.